### DISCUSSION TITLE
Can't find a working Python installation

### DISCUSSION THREAD

**Discussion by khoipkv** [gem5/gem5#1539] [Q&A]
Hello,
Can anyone help point me in the right direction on how to fix this error? Or what might have caused it?

![image](https://github.com/user-attachments/assets/84f7729d-ed0c-48a8-9d92-19963d13cecf)

This is someone else's work, and it works on their computer. We have different Python versions, so I have tried changing my Python to match theirs (3.6.9), but that also didn't work. 
We have the same version of scons.
When I tried running scons with the original gem5 (the one pulled from this repository), it worked fine. Not with this person's gem5.

This is running on Fedora 40.

Thanks a lot.

**Comment by Harshil2107** [comment]
Hi, my first guess would be that maybe the `PATH` for the updated python version is not set? or scons is not able to find python at the specified path.

**Comment by LujhCoconut** [comment]
Maybe you can try `/usr/bin/env python3 $(which scons) build/ARM/gem5.opt`

**Reply by SerenityLJY** [reply]
thanku！！

**Comment by khoipkv** [comment]
Hi, thanks for all your suggestions. I'm running just "scons" in most of my screenshots so it's faster to test the solution.

@LujhCoconut 
I have tried that command, but that didn't work as well.
![image](https://github.com/user-attachments/assets/10e7fae2-2fff-4534-85a3-64aa9806c385)

@Harshil2107 
I believe the updated version of Python is in my PATH:
![image](https://github.com/user-attachments/assets/b60d9429-3f5d-46a1-bed1-cf1e17cf3e5b)

I also tried running a virtual environment with another Python version, but that also didn't seem to work.
![image](https://github.com/user-attachments/assets/40ce88ad-d3fb-422d-a371-9215b0a0c9d3)

**Comment by Gorgeous-Patrick** [comment]
Mine does not work either.

**Comment by khoipkv** [comment]
I fixed the problem by simply running gem5 in a Ubuntu 20 container

**Comment by ConvolutedDog** [comment]
This problem is easy to encounter when using virtual python environment of `anaconda/miniconda`. My solution on CentOS is as follows:

Assuming you use `conda` to create one virtual environment `gem5-venv`:
```bash
conda create -n gem5-venv python=3.9
conda activate gem5-venv
$(which pip) install SCons
```

Then you can run the following bash script (please remember to change `YourUsername` and your `gcc/g++` path):
```bash
#!/bin/bash
export LD_LIBRARY_PATH=/home/YourUsername/miniconda3/envs/gem5-venv/lib:$LD_LIBRARY_PATH # Change to your own username!

# Set common variables
CC="/opt/rh/devtoolset-10/root/bin/gcc" # Change to your own path!
CXX="/opt/rh/devtoolset-10/root/bin/g++" # Change to your own path!
SCONS_CMD="python3 $(which scons)"
BUILD_DIR="build/X86"
JOBS=32

# First, build gem5.opt
CC="$CC" CXX="$CXX" $SCONS_CMD $BUILD_DIR/gem5.opt -j $JOBS

# Only build compile_commands.json if it doesn't exist
if [ ! -f "$BUILD_DIR/compile_commands.json" ]; then
 CC="$CC" CXX="$CXX" $SCONS_CMD $BUILD_DIR/compile_commands.json
else
 echo "$BUILD_DIR/compile_commands.json already exists, skipping..."
fi
```

Using `clang/clang++` can easily cause `gem5.opt` to not be linked successfully, as mentioned in https://github.com/gem5/gem5/issues/1722, so here I specify `CC/CXX` as `gcc/g++`. At the same time, I found that if it defaults to `clang/clang++`, this issue (Can't find a working Python installation) may also occur, but I don’t have time to find out why.

Hope it helps you!

**Comment by ramwin** [comment]
I met the same problem.

**Comment by ramwin** [comment]
copy this scripts to your `<virtualenv>/bin/python-config`
```
# Source - https://stackoverflow.com/a/48946977
# Posted by András Aszódi
# Retrieved 2025-11-12, License - CC BY-SA 3.0

#!/usr/bin/env python3

"""
This python-config script was taken from a virtual environment
created by `virtualenv`.
The only change is the hash-bang line.
The user shall copy this to ${VENV}/bin during setup.
:author: unknown + AA
:date: 2018-02-23
"""

import sys
import getopt
import sysconfig

valid_opts = ['prefix', 'exec-prefix', 'includes', 'libs', 'cflags',
              'ldflags', 'help']

if sys.version_info >= (3, 2):
    valid_opts.insert(-1, 'extension-suffix')
    valid_opts.append('abiflags')
if sys.version_info >= (3, 3):
    valid_opts.append('configdir')


def exit_with_usage(code=1):
    sys.stderr.write("Usage: {0} [{1}]\n".format(
        sys.argv[0], '|'.join('--'+opt for opt in valid_opts)))
    sys.exit(code)

try:
    opts, args = getopt.getopt(sys.argv[1:], '', valid_opts)
except getopt.error:
    exit_with_usage()

if not opts:
    exit_with_usage()

pyver = sysconfig.get_config_var('VERSION')
getvar = sysconfig.get_config_var

opt_flags = [flag for (flag, val) in opts]

if '--help' in opt_flags:
    exit_with_usage(code=0)

for opt in opt_flags:
    if opt == '--prefix':
        print(sysconfig.get_config_var('prefix'))

    elif opt == '--exec-prefix':
        print(sysconfig.get_config_var('exec_prefix'))

    elif opt in ('--includes', '--cflags'):
        flags = ['-I' + sysconfig.get_path('include'),
                 '-I' + sysconfig.get_path('platinclude')]
        if opt == '--cflags':
            flags.extend(getvar('CFLAGS').split())
        print(' '.join(flags))

    elif opt in ('--libs', '--ldflags'):
        abiflags = getattr(sys, 'abiflags', '')
        libs = ['-lpython' + pyver + abiflags]
        libs += getvar('LIBS').split()
        libs += getvar('SYSLIBS').split()
        # add the prefix/lib/pythonX.Y/config dir, but only if there is no
        # shared library in prefix/lib/.
        if opt == '--ldflags':
            if not getvar('Py_ENABLE_SHARED'):
                libs.insert(0, '-L' + getvar('LIBPL'))
            if not getvar('PYTHONFRAMEWORK'):
                libs.extend(getvar('LINKFORSHARED').split())
        print(' '.join(libs))

    elif opt == '--extension-suffix':
        ext_suffix = sysconfig.get_config_var('EXT_SUFFIX')
        if ext_suffix is None:
            ext_suffix = sysconfig.get_config_var('SO')
        print(ext_suffix)

    elif opt == '--abiflags':
        if not getattr(sys, 'abiflags', None):
            exit_with_usage()
        print(sys.abiflags)

    elif opt == '--configdir':
        print(sysconfig.get_config_var('LIBPL'))

```

> https://stackoverflow.com/questions/42020937/why-pyvenv-does-not-install-python-config
